File: 054. 螺旋矩阵.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        result = []
        top = 0
        bottom = len(matrix) - 1
        left = 0
        right = len(matrix[0]) - 1
        # 每一次循环都是在画“口”字，剥洋葱
        while top <= bottom and left <= right:
            for i in range(left, right + 1):
                result.append(matrix[top][i])
            top += 1
            for i in range(top, bottom + 1):
                result.append(matrix[i][right])
            right -= 1
            # 紧急刹车：确认没有缩成一字
            if top <= bottom and left <= right:
                for i in range(right, left - 1, -1):
                    result.append(matrix[bottom][i])
                bottom -= 1
                for i in range(bottom, top - 1, -1):
                    result.append(matrix[i][left])
                left += 1
        return result

File: 054. 螺旋矩阵.go
package main

func spiralOrder(matrix [][]int) []int {
	m, n := len(matrix), len(matrix[0])
	u, d, l, r := 0, m-1, 0, n-1
	res := make([]int, 0, m*n)

	for {
		if l > r {
			break
		}
		for j := l; j <= r; j++ {
			res = append(res, matrix[u][j])
		}
		u++

		if u > d {
			break
		}
		for i := u; i <= d; i++ {
			res = append(res, matrix[i][r])
		}
		r--

		if l > r {
			break
		}
		for j := r; j >= l; j-- {
			res = append(res, matrix[d][j])
		}
		d--

		if u > d {
			break
		}
		if l <= r {
			for i := d; i >= u; i-- {
				res = append(res, matrix[i][l])
			}
			l++
		}
	}
	return res
}
